go/services/connector/main.go

```diff
@@ -0,0 +1,620 @@
+			return nil, fmt.Errorf("failed to watch mount path: %w", err)
+		}
+	}
+
+	observation := &DeviceEndpointCreateObservation{
+		mountPath:    mountPath,
+		deviceChan:   fileMountMap.createDeviceTx,
+		fileMountMap: fileMountMap,
+		ctx:          observeCtx,
+		cancel:       cancel,
+		pollInterval: pollInterval,
+		connMonitor:  connMonitor,
+		logger:       logger,
+		watcher:      watcher,
+	}
+
+	// Initialize with existing devices
+	_, err = os.Stat(mountPath)
+	switch {
+	case err == nil:
+		// Mount path exists, check for devices
+		devices, err := getDeviceEndpointNames(mountPath)
+		if err != nil {
+			cancel()
+			watcher.Close()
+			return nil, err
+		}
+
+		fileMountMap.UpdateDeviceEndpoints(devices)
+		connMonitor.SetConnected()
+	case os.IsNotExist(err):
+		// Mount path doesn't exist yet
+		connMonitor.SetDisconnected()
+	default:
+		cancel()
+		watcher.Close()
+		return nil, &Error{Message: "Failed to access mount path", Cause: err}
+	}
+
+	// Start watcher loop
+	go observation.watchLoop()
+
+	return observation, nil
+}
+
+// watchLoop processes fsnotify events and updates the file mount map.
+func (o *DeviceEndpointCreateObservation) watchLoop() {
+	defer o.watcher.Close()
+
+	for {
+		select {
+		case <-o.ctx.Done():
+			return
+		case event, ok := <-o.watcher.Events:
+			if !ok {
+				return
+			}
+
+			// Handle creation or removal of the mount path itself.
+			if event.Name == o.mountPath && (event.Op&fsnotify.Create) != 0 {
+				// Mount appeared – start watching it.
+				_ = o.watcher.Add(o.mountPath)
+				o.connMonitor.SetConnected()
+			}
+			if event.Name == o.mountPath &&
+				(event.Op&(fsnotify.Remove|fsnotify.Rename)) != 0 {
+				// Mount disappeared
+				_ = o.watcher.Remove(o.mountPath)
+				o.connMonitor.SetDisconnected()
+				o.lastDevices = nil
+				continue
+			}
+
+			// Only process events within the mount directory when connected.
+			if !o.connMonitor.IsConnected() {
+				continue
+			}
+
+			// For any relevant operation inside mount path, reconcile.
+			if strings.HasPrefix(event.Name, o.mountPath) &&
+				(event.Op&(fsnotify.Create|fsnotify.Write|fsnotify.Remove|fsnotify.Rename)) != 0 {
+				o.reconcile()
+			}
+
+		case err, ok := <-o.watcher.Errors:
+			if ok {
+				o.logger.Warn("watcher error", "error", err)
+			}
+		}
+	}
+}
+
+// reconcile reads current state and updates the FileMountMap.
+func (o *DeviceEndpointCreateObservation) reconcile() {
+	devices, err := getDeviceEndpointNames(o.mountPath)
+	if err != nil {
+		o.logger.Warn(
+			"Failed to get device endpoint names",
+			"error",
+			err,
+		)
+		return
+	}
+
+	if !reflect.DeepEqual(devices, o.lastDevices) {
+		o.fileMountMap.UpdateDeviceEndpoints(devices)
+		o.lastDevices = devices
+	}
+}
+
+// RecvNotification receives the next device endpoint notification.
+func (o *DeviceEndpointCreateObservation) RecvNotification(
+	ctx context.Context,
+) (*DeviceNotification, error) {
+	select {
+	case notification, ok := <-o.deviceChan:
```

Comment:
Worth noting that this value is not broadcast - only a single listener will receive any given notification. That may be fine, but it's different than the way your ConnectionMonitor behaves, so I figured it was worth calling out.